frame/proxy/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+		+ GetDispatchInfo + From<frame_system::Call<Self>> + IsSubType<Module<Self>, Self>;
+
+	/// The currency mechanism.
+	type Currency: ReservableCurrency<Self::AccountId>;
+
+	/// Is a given call compatible with the proxying subsystem?
+	type IsCallable: Filter<<Self as Trait>::Call>;
+
+	/// A kind of proxy; specified with the proxy and passed in to the `IsProxyable` fitler.
+	/// The instance filter determines whether a given call may be proxied under this type.
+	type ProxyType: Parameter + Member + Ord + PartialOrd + InstanceFilter<<Self as Trait>::Call>
+		+ Default;
+
+	/// The base amount of currency needed to reserve for creating a proxy.
+	///
+	/// This is held for an additional storage item whose value size is
+	/// `sizeof(Balance)` bytes and whose key size is `sizeof(AccountId)` bytes.
+	type ProxyDepositBase: Get<BalanceOf<Self>>;
+
+	/// The amount of currency needed per proxy added.
+	///
+	/// This is held for adding 32 bytes plus an instance of `ProxyType` more into a pre-existing
+	/// storage value.
+	type ProxyDepositFactor: Get<BalanceOf<Self>>;
+
+	/// The maximum amount of proxies allowed for a single account.
+	type MaxProxies: Get<u16>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Proxy {
+		/// The set of account proxies. Maps the account which has delegated to the accounts
+		/// which are being delegated to, together with the amount held on deposit.
+		pub Proxies: map hasher(twox_64_concat) T::AccountId => (Vec<(T::AccountId, T::ProxyType)>, BalanceOf<T>);
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// There are too many proxies registered.
+		TooMany,
+		/// Proxy registration not found.
+		NotFound,
+		/// Sender is not a proxy of the account to be proxied.
+		NotProxy,
+		/// A call with a `false` `IsCallable` filter was attempted.
+		Uncallable,
+		/// A call which is incompatible with the proxy type's filter was attempted.
+		Unproxyable,
+		/// Account is already a proxy.
+		Duplicate,
+		/// Call may not be made by proxy because it may escalate its privileges.
+		NoPermission,
+	}
+}
+
+decl_event! {
+	/// Events type.
+	pub enum Event {
+		/// A proxy was executed correctly, with the given result.
+		ProxyExecuted(DispatchResult),
+		/// A call with a `false` IsCallable filter was attempted.
+		Uncallable(u32),
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		/// Deposit one of this module's events by using the default implementation.
+		fn deposit_event() = default;
+
+		/// Dispatch the given `call` from an account that the sender is authorised for through
+		/// `add_proxy`.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// Parameters:
+		/// - `real`: The account that the proxy will make a call on behalf of.
+		/// - `force_proxy_type`: Specify the exact proxy type to be used and checked for this call.
```

Comment:
with just reading the documents of these two functions, I am not sure why we have this param. 

If I call `set_proxy` and give an account to proxy me with some `ProxyType::JustTransfer`, then I probably don't want my proxy to proxy me with any other call, whilst `force_proxy_type ` allows them to override this?

Reply:
Form the code, it seems that this must be equal to the proxy_type stored in storage. In this case I think the prefix `force_` is a bit confusing.

Reply:
it must be equal to *one of* the proxy types registered. there may be several. however, in the case that there's only one which i consider to be a fairly common use case (or if you just don't care) then it may be omitted.